src/pe/delay_import.rs

```diff
@@ -0,0 +1,529 @@
+        };
+
+        let func = DelayImportFunction { name, ordinal };
+
+        Ok((func, offset))
+    }
+}
+
+#[derive(Debug)]
+pub struct DelayImportDll<'a> {
+    pub descriptor: DelayImportDescriptor,
+    ctx: PeCtx<'a>,
+    bytes: &'a [u8],
+}
+
+impl<'a> DelayImportDll<'a> {
+    pub fn functions(
+        &self,
+        sections: &'a [section_table::SectionTable],
+    ) -> DelayImportFunctionIterator<'a> {
+        // Replace sections with an actual sections ref
+        let pectx = PeCtx::new(
+            self.ctx.container,
+            self.ctx.le,
+            &sections,
+            self.ctx.file_alignment,
+            self.ctx.opts,
+            self.ctx.bytes,
+        );
+        DelayImportFunctionIterator {
+            ctx: pectx,
+            bytes: self.bytes,
+            offset: 0,
+            descriptor: self.descriptor,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct DelayImportDllIterator<'a> {
+    ctx: PeCtx<'a>,
+    bytes: &'a [u8],
+    offset: usize,
+}
+
+impl<'a> Iterator for DelayImportDllIterator<'a> {
+    type Item = error::Result<DelayImportDll<'a>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.bytes.len() {
+            return None;
+        }
+
+        let descriptor = self
+            .bytes
+            .gread_with::<DelayImportDescriptor>(&mut self.offset, scroll::LE)
+            .ok()?;
+        if descriptor.is_null() || !descriptor.is_possibly_valid() {
+            self.bytes = &[];
+            return None;
+        }
+
+        let name_table_rva = descriptor.name_table_rva;
+        let name_table_offset = utils::find_offset(
+            name_table_rva as usize,
+            self.ctx.sections,
+            self.ctx.file_alignment,
+            &self.ctx.opts,
+        )?;
+        if self.ctx.bytes.len() < name_table_offset {
+            self.bytes = &[];
+            return None;
+        }
+
+        let res = DelayImportDll {
+            descriptor,
+            ctx: self.ctx,
+            bytes: &self.ctx.bytes[name_table_offset..],
+        };
+
+        Some(Ok(res))
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct DelayImportFunctionIterator<'a> {
+    ctx: PeCtx<'a>,
+    bytes: &'a [u8],
+    offset: usize,
+    descriptor: DelayImportDescriptor,
+}
+
+impl<'a> Iterator for DelayImportFunctionIterator<'a> {
+    type Item = error::Result<DelayImportFunction<'a>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.bytes.len() {
+            return None;
+        }
+
+        let ctx = Ctx::new(self.ctx.container, scroll::LE);
+        let thunk = self
+            .bytes
+            .gread_with::<DelayImportThunk>(&mut self.offset, ctx)
+            .ok()?;
+        if thunk.is_null() {
+            self.bytes = &[];
+            return None;
+        }
+
+        let ctx = (self.descriptor, self.ctx);
+        let func = self.bytes.pread_with::<DelayImportFunction>(0, ctx).ok()?;
```

Comment:
all control flow appears to do `ok()?` which converts Error -> Option, so the iteration item can just be `type Item = DelayImportFunction<'a>` if that's your intention (to drop all error messages during iteration, and stop on first error)